File: lib/models/category.model.dart
import 'product.model.dart';

class Category {
  final String id;
  final String name;
  final String desc;
  final int ordering;
  final String status;
  // final List<Product> products;
  final String? imageUrl;
  final String slug;

  Category({
    required this.id,
    required this.name,
    this.desc = '',
    this.ordering = 0,
    this.status = 'active',
    // required this.products,
    this.imageUrl,
    required this.slug,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    // var productList = json['products'] as List? ?? [];
    // List<Product> productsList = productList.map((productMap) {
    //   return Product.fromJson(productMap as Map<String, dynamic>);
    // }).toList();

    return Category(
      id: json['_id'],
      name: json['name'],
      desc: json['desc'] ?? '',
      ordering: json['ordering']?.toInt() ?? 0,
      status: json['status'] ?? 'active',
      // products: productsList,
      imageUrl: json['image_url'],
      slug: json['slug'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'desc': desc,
      'ordering': ordering,
      'status': status,
      // 'products': products.map((product) => product.toJson()).toList(),
      'image_url': imageUrl,
      'slug': slug,
    };
  }

  Category copyWith({
    String? id,
    String? name,
    String? desc,
    int? ordering,
    String? status,
    List<Product>? products,
    String? imageUrl,
    String? slug,
  }) {
    return Category(
      id: id ?? this.id,
      name: name ?? this.name,
      desc: desc ?? this.desc,
      ordering: ordering ?? this.ordering,
      status: status ?? this.status,
      // products: products ?? this.products,
      imageUrl: imageUrl ?? this.imageUrl,
      slug: slug ?? this.slug,
    );
  }
}

File: src/v1/models/category.model.js
const { Schema, model } = require('mongoose');
var slugify = require('slugify')

const COLLECTION_NAME = 'Categories';
const DocumentName = 'Category';

const categorySchema = new Schema({
    name: { type: String, required: true },
    desc: { type: String },
    ordering: { type: Number, default: 0 },
    status: { type: String, enum: ['active', 'inactive'], default: 'active' },
    products: [{ type: Schema.Types.ObjectId, ref: 'Product' }],
    image_url: { type: String },
    slug: { type: String, unique: true }
}, {
    timestamps: true,
    collection: COLLECTION_NAME
});

// Tạo index để tăng hiệu suất tìm kiếm.
// categorySchema.index({ name: 1, slug: 1 }, { unique: true });
categorySchema.pre('save', function (next) {
    this.slug = slugify(this.name, { lower: true });
    next();
});

module.exports = model(DocumentName, categorySchema);
